### ISSUE TITLE
Use includes?

### ISSUE THREAD

**Issue by mrdoob** [erichlof/THREE.js-PathTracing-Renderer#6]
As far as I can see, the demos have a lot of repeated shader code.

How about doing this something like this?

```html
<script src="three.js"></script>
<script src="three-pathtracing.js"></script>
```
`three-pathtracing.js` would have this inside:

```js
THREE.ShaderChunk[ 'pathtracing_functions' ] = `
//-----------------------------------------------------------------------
float SphereIntersect( float rad, vec3 pos, Ray r )
//-----------------------------------------------------------------------
{
	vec3 op = pos - r.origin;
	float b = dot(op, r.direction);
	float det = b * b - dot(op,op) + rad * rad;
       	if (det < 0.0)
		return INFINITY;
        
	det = sqrt(det);	
	float t1 = b - det;
	if( t1 > 0.0 )
		return t1;
		
	float t2 = b + det;
	if( t2 > 0.0 )
		return t2;
	return INFINITY;	
}
// etc
`;
```

Then all the demos can include that code by adding this in the glsl:

```glsl
#include <pathtracing_functions>
```

**Comment by erichlof** [comment]
Hi @mrdoob ,
What a great idea!  Yes, as the number of demos has grown,  I was thinking that there is a lot of repeated boiler-plate shader code as well as init code in the init() function (the javascript portion located later in the demo .html files).  I didn't want to have to enclose every shader code line in strings "", but from what you posted as an example, maybe I can just turn it into a THREE.ShaderChunk .

This would significantly reduce the demos code length, as well as make the main path tracing renderer more reusable, especially when I make small changes to it.

Thanks for the suggestion, I will try to get a pathtracingRenderer.js file up and running.
-Erich

**Comment by erichlof** [comment]
Hi @mrdoob ,
Just a follow up - I took your suggestion to heart and refactored the pathtracing engine functions using a more modular approach with #includes.
I have updated the Cornell box demo as the first demo to use this new code base:
[Cornell Box Demo using #includes](https://erichlof.github.io/THREE.js-PathTracing-Renderer/ThreeJS_PathTracing_Renderer_CornellBox_DirectLighting.html)

and here is the new source code:
[Demo source code](https://github.com/erichlof/THREE.js-PathTracing-Renderer/blob/gh-pages/ThreeJS_PathTracing_Renderer_CornellBox_DirectLighting.html)
and the file with common pathtracing functions, setup uniforms, and defines:
[common pathtracing functions](https://github.com/erichlof/THREE.js-PathTracing-Renderer/blob/gh-pages/js/pathTracingCommon.js)

Thanks again for the suggestion!  It cut down the amount of demo .html code significantly.  Plus, it's nice having the common pathtracing utility functions all located in 1 file.  This will make small changes and maintenance much easier.  Now, I have to convert all the other demo .html files 1 by 1, but the updates will be worth it later on down the line.

p.s. Can't wait for WebGL 2.0 support in three.js (it will allow me to add robust LBVH acceleration structures using bit manipulations and Morton codes in the WebGL shaders, all on the GPU!) :-)

Take care,
-Erich
